src/SMPProducerConsumer.java

```diff
@@ -0,0 +1,190 @@
+                int preferredPerson = findPreferredPerson(freePerson);
+
+                synchronized (assignment) {
+                    int pairedPerson = proposeToPerson(freePerson, preferredPerson);
+                    if (pairedPerson < 0) {
+                        assign(freePerson, preferredPerson);
+                    } else {
+                        int rejected;
+                        if (prefers(preferredPerson, freePerson, pairedPerson)) {
+                            rejected = freePerson;
+                        } else {
+                            rejected = pairedPerson;
+                            assign(freePerson, preferredPerson);
+                        }
+                        _queue.add(rejected);
+                    }
+                }
+            }
+        }
+
+        private boolean prefers(int preferredPerson, int freePerson, int pairedPerson) {
+            int rankFreePerson = -1;
+            int rankPairedPerson = -1;
+            for (int i = 0; i < pref2.length; i++) {
+                if (pref2[preferredPerson][i] - 1 == freePerson) {
+                    rankFreePerson = i;
+                }
+                if (pref2[preferredPerson][i] - 1 == pairedPerson) {
+                    rankPairedPerson = i;
+                }
+            }
+
+            return rankFreePerson > rankPairedPerson;
+        }
+
+        private void assign(int freePerson, int preferredPerson) {
+            for (int i = 0; i < assignment.length; i++) {
+                assignment[i][preferredPerson] = i == freePerson;
+            }
+        }
+
+        private int findPreferredPerson(int freePerson) {
+            return pref1[freePerson][proposed[freePerson]] - 1;
+        }
+
+        private int proposeToPerson(int freePerson, int preferredPerson) {
+            int pairedWith = -1;
+            for (int i = 0; i < assignment.length; i++) {
+                if (assignment[i][preferredPerson]) {
+                    pairedWith = i;
+                    break;
+                }
+            }
+            proposed[freePerson]++;
+            return pairedWith;
+        }
+
+
+    }
+
+    public String run() {
+        int count = pref1.length;
+        List<Integer> allFree = IntStream.range(0, count).boxed().collect(Collectors.toList());
+        ConcurrentLinkedQueue<Integer> freeList = new ConcurrentLinkedQueue<>(allFree);
+
+        ArrayList<Future> waitFor = new ArrayList<>();
+
+        ExecutorService pool = Executors.newCachedThreadPool();
+
+        for (int i = 0; i < count; i++) {
+            waitFor.add(pool.submit(new Worker(freeList)));
+        }
+
+        for (Future w : waitFor) {
+            try {
+                w.get();
+            } catch (InterruptedException e) {
+                e.printStackTrace();
+            } catch (ExecutionException e) {
+                e.printStackTrace();
+            }
+        }
+
+        pool.shutdown();
```

Comment:
This algorithm actually terminates, so it waits for all the threads to finish and then shutdowns the thread pool.